system/src/system_power_manager.cpp

```diff
@@ -477,4 +495,9 @@ void PowerManager::deinit() {
   detachInterrupt(LOW_BAT_UC);
 }
 
+void PowerManager::usbStateChangeHandler(HAL_USB_State state, void* context) {
+  PowerManager* power = (PowerManager*)context;
+  power->update();
```

Comment:
I'm not sure that this is enough. When I was testing the previous iteration of this PR, I didn't see DPDM detection automatically running on USB cable disconnection/reconnection. Could you please verify that it does happen and if not, we need to set a flag to force it in `handleUpdate()`.

Reply:
Did you connect Boron to the USB host? If so, it wouldn't change the charging current. `if (HAL_USB_Get_State() == HAL_USB_STATE_DETACHED) `

---
I added the log in:
```
...
  power_source_t src = g_powerSource;
  if (pwr_good) {
    uint8_t vbus_stat = status >> 6;
    LOG_DEBUG(INFO, "vbus_stat: 0x%x, usb state: %d", vbus_stat, HAL_USB_Get_State());
    switch (vbus_stat) {
      case 0x01:
...
```
Keep VIN pin connected and disconnect/reconnect USB cable, got the log:
```
00124238 [sys.power] TRACE: re-enabling charging
0000149072 [sys.power] TRACE: Battery state UNKNOWN -> CHARGING
0000149073 [sys.power] INFO: vbus_stat: 0x1, USB state: 3
0000149076 [sys.power] INFO: vbus_stat: 0x1, USB state: 3
0000158376 [sys.power] TRACE: Battery state CHARGING -> DISCHARGING
0000158644 [sys.power] TRACE: Battery state DISCHARGING -> NOT_CHARGING
0000158645 [sys.power] INFO: vbus_stat: 0x0, USB state: 3
0000158854 [sys.power] TRACE: Battery state NOT_CHARGING -> CHARGING
0000158854 [sys.power] INFO: vbus_stat: 0x1, USB state: 3
0000168480 [sys.power] INFO: vbus_stat: 0x1, USB state: 0
0000172427 [sys.power] INFO: vbus_stat: 0x1, USB state: 3
0000172430 [sys.power] INFO: vbus_stat: 0x1, USB state: 3
0000176617 [sys.power] TRACE: Battery state CHARGING -> DISCHARGING
0000176886 [sys.power] TRACE: Battery state DISCHARGING -> NOT_CHARGING
0000176887 [sys.power] INFO: vbus_stat: 0x0, USB state: 3
0000177049 [sys.power] TRACE: Battery state NOT_CHARGING -> CHARGING
0000177049 [sys.power] INFO: vbus_stat: 0x1, USB state: 3
0000183957 [sys.power] INFO: vbus_stat: 0x1, USB state: 0
0000186186 [sys.power] INFO: vbus_stat: 0x1, USB state: 3
0000186189 [sys.power] INFO: vbus_stat: 0x1, USB state: 3
0000192023 [sys.power] INFO: vbus_stat: 0x1, USB state: 0
```
Number 3(`HAL_USB_STATE_DEFAULT`) means it connected to the USB host. When disconnecting USB host, the charging current increased to 800mA.

Reply:
Does it get back to 500mA though? And is it possible that these changes are actually caused by the battery state fluctuation we see here in the log? What happens if you connect a USB adapter?

Reply:
> Does it get back to 500mA though? 
1. Insert USB current measure tool to a socket, use a USB cable to connect the tool and  Boron VIN pin, the current on the USB current measure tool is 800mA.
2. Connect USB Host to the USB port on Boron, the current decreases to less than 500mA.
3. Disconnect USB Host, the current increases to 800mA.

> And is it possible that these changes are actually caused by the battery state fluctuation we see here in the log? 
The log is triggered by the interrupts of VIN pin and USB. In the interrupt handler, `update_` flag is updated, the log is printed in the thread loop.

> What happens if you connect a USB adapter?
Since the VIN pin is connected to a USB adapter, I didn't try to connect the USB port to a USB adapter. I can try it tomorrow to see whether the charging current will remain high.